File: src/ReactNativeSunmiBarcodeScannerModule.ts
import { NativeModule, requireNativeModule } from "expo";
import { Platform } from "react-native";

import { ReactNativeSunmiBarcodeScannerModuleEvents } from "./ReactNativeSunmiBarcodeScanner.types";

export type ScannerOperationMode = "ON_DEMAND" | "CONTINUOUS";
export type ScannerPriority =
  | "PREFER_USB"
  | "PREFER_SERIAL"
  | "USB_ONLY"
  | "SERIAL_ONLY";
export type ScannerType = "USB" | "SERIAL" | "BOTH" | "NONE";
export type DataDistributeType =
  | "TYPE_KEYBOARD"
  | "TYPE_BROADCAST"
  | "TYPE_KEYBOARD_AND_BROADCAST";

export interface ScannerInfo {
  type: ScannerType;
  isConnected: boolean;
  deviceName: string | null;
  pid: number | null;
  vid: number | null;
}

export interface UsbDeviceInfo {
  deviceName: string | null;
  vendorId: number;
  productId: number;
  pidVidKey: string;
  deviceClass: number;
  deviceSubclass: number;
  deviceProtocol: number;
  interfaceCount: number;
  isCompatible: boolean;
  hasPermission?: boolean;
  interfaces?: UsbInterfaceInfo[];
}

export interface UsbInterfaceInfo {
  id: number;
  interfaceClass: number;
  interfaceSubclass: number;
  interfaceProtocol: number;
  endpointCount: number;
  name: string;
}

class PlatformNotSupportedError extends Error {
  constructor() {
    super(
      "Platform not supported. Sunmi barcode scanner is only available on Android."
    );
    this.name = "PlatformNotSupportedError";
  }
}

const checkPlatform = () => {
  if (Platform.OS !== "android") {
    throw new PlatformNotSupportedError();
  }
};

declare class ReactNativeSunmiBarcodeScannerModuleNative extends NativeModule<ReactNativeSunmiBarcodeScannerModuleEvents> {
  initializeScanner(): void;
  setScannerOperationMode(mode: ScannerOperationMode): void;
  getScannerOperationMode(): ScannerOperationMode;
  setScannerPriority(priority: ScannerPriority): void;
  getScannerPriority(): ScannerPriority;
  getAvailableScanners(): Promise<ScannerInfo[]>;
  getCurrentScannerType(): ScannerType;
  getOptimalScannerType(): ScannerType;
  setScanTimeout(timeout: number): void;
  setBeep(enabled: boolean): void;
  setToast(enabled: boolean): void;
  getToast(): boolean;
  setUsbScannerMode(mode: number): void;
  setDataDistributeType(type: DataDistributeType): void;
  getAllUsbDevices(): Promise<UsbDeviceInfo[]>;
  addCompatibleUsbScanner(productId: number, vendorId: number): boolean;
  removeCompatibleUsbScanner(productId: number, vendorId: number): boolean;
  getCompatibleUsbScanners(): string[];
  resetCompatibleUsbScanners(): void;
  requestUsbPermission(vendorId: number, productId: number): boolean;
  testUsbScannerModes(vendorId: number, productId: number): void;
  scanQRCode: () => Promise<string>;
  cancelScan: () => Promise<void>;
}

const nativeModule =
  Platform.OS === "android"
    ? requireNativeModule<ReactNativeSunmiBarcodeScannerModuleNative>(
        "ReactNativeSunmiBarcodeScanner"
      )
    : null;

// Platform-aware wrapper
const ReactNativeSunmiBarcodeScannerModule = {
  initializeScanner(): void {
    checkPlatform();
    nativeModule!.initializeScanner();
  },

  setScannerOperationMode(mode: ScannerOperationMode): void {
    checkPlatform();
    nativeModule!.setScannerOperationMode(mode);
  },

  getScannerOperationMode(): ScannerOperationMode {
    checkPlatform();
    return nativeModule!.getScannerOperationMode();
  },

  setScannerPriority(priority: ScannerPriority): void {
    checkPlatform();
    nativeModule!.setScannerPriority(priority);
  },

  getScannerPriority(): ScannerPriority {
    checkPlatform();
    return nativeModule!.getScannerPriority();
  },

  async getAvailableScanners(): Promise<ScannerInfo[]> {
    checkPlatform();
    return nativeModule!.getAvailableScanners();
  },

  getCurrentScannerType(): ScannerType {
    checkPlatform();
    return nativeModule!.getCurrentScannerType();
  },

  getOptimalScannerType(): ScannerType {
    checkPlatform();
    return nativeModule!.getOptimalScannerType();
  },

  setScanTimeout(timeout: number): void {
    checkPlatform();
    nativeModule!.setScanTimeout(timeout);
  },

  setBeep(enabled: boolean): void {
    checkPlatform();
    nativeModule!.setBeep(enabled);
  },

  setToast(enabled: boolean): void {
    checkPlatform();
    nativeModule!.setToast(enabled);
  },

  getToast(): boolean {
    checkPlatform();
    return nativeModule!.getToast();
  },

  setUsbScannerMode(mode: number): void {
    checkPlatform();
    nativeModule!.setUsbScannerMode(mode);
  },

  setDataDistributeType(type: DataDistributeType): void {
    checkPlatform();
    nativeModule!.setDataDistributeType(type);
  },

  async getAllUsbDevices(): Promise<UsbDeviceInfo[]> {
    checkPlatform();
    return nativeModule!.getAllUsbDevices();
  },

  addCompatibleUsbScanner(productId: number, vendorId: number): boolean {
    checkPlatform();
    return nativeModule!.addCompatibleUsbScanner(productId, vendorId);
  },

  removeCompatibleUsbScanner(productId: number, vendorId: number): boolean {
    checkPlatform();
    return nativeModule!.removeCompatibleUsbScanner(productId, vendorId);
  },

  getCompatibleUsbScanners(): string[] {
    checkPlatform();
    return nativeModule!.getCompatibleUsbScanners();
  },

  resetCompatibleUsbScanners(): void {
    checkPlatform();
    nativeModule!.resetCompatibleUsbScanners();
  },

  requestUsbPermission(vendorId: number, productId: number): boolean {
    checkPlatform();
    return nativeModule!.requestUsbPermission(vendorId, productId);
  },

  testUsbScannerModes(vendorId: number, productId: number): void {
    checkPlatform();
    nativeModule!.testUsbScannerModes(vendorId, productId);
  },

  async scanQRCode(): Promise<string> {
    checkPlatform();
    return nativeModule!.scanQRCode();
  },

  async cancelScan(): Promise<void> {
    checkPlatform();
    return nativeModule!.cancelScan();
  },
};

export default ReactNativeSunmiBarcodeScannerModule;

File: android/src/main/java/expo/modules/sunmibarcodescanner/ReactNativeSunmiBarcodeScannerModule.kt
package expo.modules.sunmibarcodescanner

import android.app.Activity
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise

class ReactNativeSunmiBarcodeScannerModule : Module() {

  private var barcodeScanner = SunmiBarcodeScanner()
  private val context get() = requireNotNull(appContext.reactContext)

  // Each module class must implement the definition function. The definition consists of components
  // that describes the module's functionality and behavior.
  // See https://docs.expo.dev/modules/module-api for more details about available components.
  override fun definition() = ModuleDefinition {
    // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
    // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
    // The module will be accessible from `requireNativeModule('ReactNativeSunmiBarcodeScanner')` in JavaScript.
    Name("ReactNativeSunmiBarcodeScanner")

    // ----------------------------------------
    // Sunmi Barcode Scanner SDK public methods
    // ----------------------------------------

    Function("initializeScanner") {
      barcodeScanner.initializeScanner(context)
    }

    Function("setScannerOperationMode") { mode: String ->
      val operationMode = ScannerOperationMode.valueOf(mode)
      barcodeScanner.setScannerOperationMode(context, operationMode)
    }

    Function("getScannerOperationMode") {
      return@Function barcodeScanner.getScannerOperationMode().modeName
    }

    Function("setScannerPriority") { priority: String ->
      val scannerPriority = ScannerPriority.valueOf(priority)
      barcodeScanner.setScannerPriority(context, scannerPriority)
    }

    Function("getScannerPriority") {
      return@Function barcodeScanner.getScannerPriority().priorityName
    }

    AsyncFunction("getAvailableScanners") { promise: Promise ->
      try {
        val scanners = barcodeScanner.getAvailableScanners(context)
        val scannersArray = scanners.map { scanner ->
          mapOf(
            "type" to scanner.type.typeName,
            "isConnected" to scanner.isConnected,
            "deviceName" to scanner.deviceName,
            "pid" to scanner.pid,
            "vid" to scanner.vid
          )
        }
        promise.resolve(scannersArray)
      } catch (e: Exception) {
        promise.reject("SCANNER_DETECTION_ERROR", e.message, e)
      }
    }

    Function("getCurrentScannerType") {
      return@Function barcodeScanner.getCurrentScannerType().typeName
    }

    Function("getOptimalScannerType") {
      return@Function barcodeScanner.getOptimalScannerType(context).typeName
    }

    Function("setScanTimeout") { timeout: Long ->
      barcodeScanner.setScanTimeout(timeout)
    }

    Function("setBeep") { enabled: Boolean ->
      barcodeScanner.setBeep(enabled)
    }

    Function("setToast") { enabled: Boolean ->
      barcodeScanner.setToast(enabled)
    }

    Function("getToast") {
      barcodeScanner.getToast()
    }

    Function("setUsbScannerMode") { mode: Int ->
      barcodeScanner.setUsbScannerMode(context, mode)
    }

    Function("setDataDistributeType") { type: String ->
      barcodeScanner.setDataDistributeType(context, type)
    }

    AsyncFunction("getAllUsbDevices") { promise: Promise ->
      try {
        val devices = barcodeScanner.getAllUsbDevices(context)
        promise.resolve(devices)
      } catch (e: Exception) {
        promise.reject("USB_DEVICE_ERROR", e.message, e)
      }
    }

    Function("addCompatibleUsbScanner") { productId: Int, vendorId: Int ->
      barcodeScanner.addCompatibleUsbScanner(productId, vendorId)
    }

    Function("removeCompatibleUsbScanner") { productId: Int, vendorId: Int ->
      barcodeScanner.removeCompatibleUsbScanner(productId, vendorId)
    }

    Function("getCompatibleUsbScanners") {
      barcodeScanner.getCompatibleUsbScanners()
    }

    Function("resetCompatibleUsbScanners") {
      barcodeScanner.resetCompatibleUsbScanners()
    }

    // USB Troubleshooting methods
    Function("requestUsbPermission") { vendorId: Int, productId: Int ->
      barcodeScanner.requestUsbPermission(context, vendorId, productId)
    }

    Function("testUsbScannerModes") { vendorId: Int, productId: Int ->
      barcodeScanner.testUsbScannerModes(context, vendorId, productId)
    }

    AsyncFunction("scanQRCode")  { promise: Promise ->
      barcodeScanner.scanQRCode(context, promise)
    }

    AsyncFunction("cancelScan")  { promise: Promise ->
      barcodeScanner.cancelScan(context, promise)
    }

    // Cleanup when module is destroyed
    OnDestroy {
      barcodeScanner.destroy()
    }
  }
}
